pkg/reconciler/v1alpha1/autoscaling/kpa/resources/metric.go

```diff
@@ -18,19 +18,21 @@ package resources
 
 import (
 	"context"
+	"time"
 
 	"github.com/knative/pkg/logging"
 	"github.com/knative/serving/pkg/apis/autoscaling/v1alpha1"
 	"github.com/knative/serving/pkg/autoscaler"
+	autoscalerConfig "github.com/knative/serving/pkg/autoscaler/config"
 )
 
 // MakeMetric constructs a Metric resource from a PodAutoscaler taking
 // into account the PA's ContainerConcurrency and the relevant
 // autoscaling annotation.
-func MakeMetric(ctx context.Context, pa *v1alpha1.PodAutoscaler, config *autoscaler.Config) *autoscaler.Metric {
+func MakeMetric(ctx context.Context, pa *v1alpha1.PodAutoscaler, config *autoscalerConfig.Config) *autoscaler.Metric {
 	logger := logging.FromContext(ctx)
 
-	target := config.TargetConcurrency(pa.Spec.ContainerConcurrency)
+	target := config.TargetConcurrency(float64(pa.Spec.ContainerConcurrency))
```

Comment:
Shall we use `pa.Target()` instead of `pa.Spec.ContainerConcurrency`? BTW is it still necessary to call `config.TargetConcurrency` to apply `container-concurrency-target-percentage`. It looks complicated to me that we have both `target` and `container-concurrency-target-percentage`. `target` times `container-concurrency-target-percentage` is the actual value to use.

Reply:
The parameters, methods and call graph used to calculate the effective target concurrency is really confusing!  I traced this out and I think I can simplify it as such:

![image](https://user-images.githubusercontent.com/1103629/53881794-d646d280-4014-11e9-8175-a5b1dc6aac26.png)

We have to take a couple things into account when setting a target:
1. the user-specified concurrency limitation of the container
2. the cluster default target
3. the target annotation on the pod autoscaler

Previously cluster defaults were applied when the metric is created in memory (in the autoscaler).  Now cluster defaults are applied when the pod autoscaler is created by the revision, via the webhook.  So without backward compatibility, we would only need to consider (1) and (3).  However,  preexisting pod autoscalers may not have the expected annotations, so we should fall back on applying the cluster defaults (2).

There is only one place where we call `TargetConcurrency` so we can just move that code into the `MakeMetric` method and all the calculation will be in one place.

@yanweiguo what do you think?

Reply:
(I'll show you by updating this pull request)

Reply:
I can do this separately: https://github.com/knative/serving/pull/3373